File: semester 4/MPP/C#FTS/C#FTS/Domain/Show.cs
using System;
using System.Data;

namespace C_FTS.Domain
{
    public class Show : Entity<int>
    {
        public DateTime Date { get; set; }
        public string Title { get; }
        public int SoldSeats { get; private set; }
        public Venue Venue { get; }

        public Show(int id, DateTime date, string title, int soldSeats, Venue venue)
            : base(id)
        {
            Date = date;
            Title = title ?? throw new ArgumentNullException(nameof(title));
            SoldSeats = soldSeats;
            Venue = venue ?? throw new ArgumentNullException(nameof(venue));
        }

        public Show(IDataReader reader, Venue venue)
            : base(reader.GetInt32(reader.GetOrdinal("id")))
        {
            Date = reader.GetDateTime(reader.GetOrdinal("date"));
            Title = reader.GetString(reader.GetOrdinal("title"));
            SoldSeats = reader.GetInt32(reader.GetOrdinal("sold_seats"));
            Venue = venue ?? throw new ArgumentNullException(nameof(venue));
        }

        public bool IsSoldOut()
        {
            return Venue.Capacity <= SoldSeats;
        }

        // This function always sells the ticket without checking capacity
        public Ticket SellTicket(string buyerName, int numberOfSeats)
        {
            return new Ticket(
                buyerName,
                numberOfSeats,
                DateTime.Now,
                this
            );
        }

        public void AddToSoldSeats(int numberOfSoldSeats)
        {
            this.SoldSeats += numberOfSoldSeats;
        }

        public string FormattedDate => Date.ToString("dd/MM/yyyy");
        public string FormattedTime => Date.ToString("HH:mm");
        public int RemainingSeats => Venue.Capacity - SoldSeats;
    }
}

File: semester 4/MPP/JFTS/FestivalDomain/src/main/java/ro/mpp/domain/Show.java
package ro.mpp.domain;

import jakarta.persistence.*;
import lombok.AccessLevel;
import lombok.Getter;
import lombok.NoArgsConstructor;

import java.sql.ResultSet;
import java.sql.SQLException;
import java.time.LocalDateTime;

@Getter
@Entity
@Table(name = "show")
@NoArgsConstructor(access = AccessLevel.PROTECTED)
public class Show extends HasId<Integer> {
    @Column(name = "date")
    private String date;

    @Column(name = "title")
    private String title;

    @Column(name = "sold_seats")
    private Integer soldSeats;

    @ManyToOne(fetch = FetchType.EAGER)
    @JoinColumn(name = "venue_id")
    private Venue venue;

    public Show(Integer id, String date, String title, Integer soldSeats, Venue venue) {
        super(id);
        this.date = date;
        this.title = title;
        this.soldSeats = soldSeats;
        this.venue = venue;
    }

    public Show(ResultSet rs, Venue venue) throws SQLException {
        super(rs.getInt("id"));
        this.date = rs.getString("date");
        this.title = rs.getString("title");
        this.soldSeats = rs.getInt("sold_seats");
        this.venue = venue;
    }

    public boolean isSoldOut() {
        return this.venue.getCapacity() <= this.soldSeats;
    }

    public Integer remainingSeats() {
        return this.venue.getCapacity() - this.soldSeats;
    }

    public Ticket sellTicket(String buyerName, Integer numberOfSeats) {
        return new Ticket(
                buyerName,
                numberOfSeats,
                LocalDateTime.now().toString(),
                this
        );
    }

    public void addToSoldSeats(int numberOfSeats) {
        this.soldSeats += numberOfSeats;
    }
}
